### DISCUSSION TITLE
Mailparse and chumsky questions

### DISCUSSION THREAD

**Discussion by soywod** [stalwartlabs/mail-parser#36] [General]
1. Do you know about the [mailparse](https://github.com/staktrace/mailparse) crate? What were your motivation when you started this repo (instead of using mailparse)?
2. I checked your code and it looks really nice (well done :muscle:), but I find hand-made state machines a bit hard to understand (for new comers on your codebase). I recently open [a PR for the lettre](https://github.com/lettre/lettre/pull/839) crate in order to refactor their hand-made mailbox parsing using parsers combinators ([chumsky](https://github.com/zesterer/chumsky)). IMHO parsers combinators fit so well RFCs: you just need to implement small parsers and combine them together. I also find the result much readable and maintainable. Plus the perf loss is not that big. Is it sth you would be interested in (to receive PR using parsers combinators)?

**Comment by mdecimus** [comment]
> Do you know about the [mailparse](https://github.com/staktrace/mailparse) crate? What were your motivation when you started this repo (instead of using mailparse)?

Yes, but I wanted to implement something more robust from scratch. Mailparse does not support parsing malformed messages and can handle only a limited number of encodings.

> I checked your code and it looks really nice (well done 💪), but I find hand-made state machines a bit hard to understand (for new comers on your codebase). I recently open https://github.com/lettre/lettre/pull/839 crate in order to refactor their hand-made mailbox parsing using parsers combinators ([chumsky](https://github.com/zesterer/chumsky)). IMHO parsers combinators fit so well RFCs: you just need to implement small parsers and combine them together. I also find the result much readable and maintainable. Plus the perf loss is not that big. Is it sth you would be interested in (to receive PR using parsers combinators)?

Parser combinators are really nice but in my opinion they are not a good fit for e-mail because it is quite common to find messages that do not fully conform to the RFCs. For instance, many buggy mail clients generate malformed MIME headers and other clients submit messages with LF instead of CRLF (or a combination of both). These are scenarios that mailparse does not support and would be tricky to implement with a parser combinator. The main advantage of `mail-parser` is that is able to parse malformed/non-conformant e-mail messages as long as these do not deviate much from the standard.

**Reply by soywod** [reply]
> Yes, but I wanted to implement something more robust from scratch. Mailparse does not support parsing malformed messages and can handle only a limited number of encodings.

Totally makes sense, I am using `mailparse` right now and I got many issues related to parsing.

> Parser combinators are really nice but in my opinion they are not a good fit for e-mail because it is quite common to find messages that do not fully conform to the RFCs.

Well you can write different versions of parsers (some more flexible, other stricter), try to parser and do some error recovery etc. But I totally agree on the buggyness of the mail world, be able to parse malformed MIME messages is a must to have a decent user experience. Anyway thank you for having answered my questions. I will try to replace `mailparse` with `mail-parser`, I may have other questions coming soon!
